### PR TITLE
[faster-rcnn] Use executor.reshape for testing

### PR DESCRIPTION
**Pull request by sxjscience**
1. Use executor.reshape for testing.

 The original test function creates a new executor every minibatch, which is not so memory-efficient. This     PR uses `executor.reshape` to generate a new shared executor every minibatch. The inner-space of the executor will be upsized if a larger data-batch comes and no up-sizing is needed otherwise.
2. Fix format-string on Windows

 The previous format string  has some problems on Windows and this PR fixes it.
3. Add warmup

 `warmup.py` is missed for end2end training

@precedenceguo @winstywang @tornadomeet

### CODE REVIEW

File: example/rcnn/rcnn/symbol.py [hunk 1]
```diff
@@ -382,6 +382,6 @@ def get_faster_rcnn(num_classes=21, num_anchors=9):
     bbox_pred = mx.symbol.Reshape(data=bbox_loss, shape=(config.TRAIN.IMS_PER_BATCH, -1, 4 * num_classes), name='bbox_pred_reshape')
 
     # group output
-    group = mx.symbol.Group([rois[1], rpn_cls_loss, rpn_bbox_loss, cls_prob, bbox_pred])  # rois[1] is used for evaluation
+    group = mx.symbol.Group([mx.sym.BlockGrad(rois[1]), rpn_cls_loss, rpn_bbox_loss, cls_prob, bbox_pred])  # rois[1] is used for evaluation
```
Comment (tornadomeet) [inline example/rcnn/rcnn/symbol.py]: does it influence the training if not use BlockGrad here?

File: example/rcnn/rcnn/warmup.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
+from mxnet.lr_scheduler import LRScheduler
```
Comment (tornadomeet) [inline example/rcnn/rcnn/warmup.py]: in my experiments, warmup is no need, so we'd better rmove warmuplr from: https://github.com/dmlc/mxnet/blob/master/example/rcnn/train_end2end.py#L101 or bring the args of whether use warmup.

File: example/rcnn/rcnn/detector.py [hunk 3]
```diff
@@ -33,36 +33,50 @@ def im_detect(self, im_array, im_info=None, roi_array=None):
...
-                                         grad_req='null', aux_states=self.aux_params)
-        output_dict = {name: nd for name, nd in zip(self.symbol.list_outputs(), self.executor.outputs)}
-        self.executor.forward(is_train=False)
+            # execute
+            self.executor = self.symbol.bind(self.ctx, self.arg_params, args_grad=None,
+                                             grad_req='null', aux_states=self.aux_params)
+            executor = self.executor
+        else:
+            # Test whether we need to upsizing
+            if np.prod(im_array.shape) > np.prod(self.executor.arg_dict['data'].shape):
```
Comment (precedenceguo) [inline example/rcnn/rcnn/detector.py]: Actually shape of rois could also vary in the rcnn case.

File: example/rcnn/rcnn/symbol.py (hunk 1 above)
Comment (precedenceguo) [inline example/rcnn/rcnn/symbol.py]: proposal_target op gives the rois output and does not require backward dependency since no grad from the rcnn sampler (proposal_target) and rcnn branch is backproped to rpn branch. blocking rois grad or not does not affect training.

Comment (sxjscience) [inline example/rcnn/rcnn/symbol.py]: In fact I tried to run without blockgrad and met an error in the backward pass, which said that some of the symbol needs out_grad.

Comment (tornadomeet) [inline example/rcnn/rcnn/symbol.py]: @sxjscience  so if that, we can add BlockGrad() here.  (i do not met this error before)

File: example/rcnn/rcnn/detector.py (hunk 3 above)
Comment (sxjscience) [inline example/rcnn/rcnn/detector.py]: Yes, I've only tested the rpn case. Shall revise this part.

File: example/rcnn/rcnn/symbol.py (hunk 1 above)
Comment (sxjscience) [inline example/rcnn/rcnn/symbol.py]: @tornadomeet OK, it's really a strange problem and I haven't figured out the cause.

File: example/rcnn/rcnn/warmup.py (hunk 2 above)
Comment (sxjscience) [inline example/rcnn/rcnn/warmup.py]: So we should revise this line directly to `'lr_scheduler': mx.lr_scheduler.FactorScheduler(factor_step, 0.1)`?

Comment (tornadomeet) [inline example/rcnn/rcnn/warmup.py]: @sxjscience yes, i think so.   i write warmup because there is a backward grad bug(not be zero) in costum op and this will bring nan during training. now this has been fixed, so warm up is no need, and i found it is even bad for training(i forget to revise that at last pr, but i do not use it when training face detection).

File: example/rcnn/rcnn/detector.py (hunk 3 above)
Comment (sxjscience) [inline example/rcnn/rcnn/detector.py]: @precedenceguo Also, when running the original test script, I find the memory consumption will vary dramatically between around 2G to 7G and may occasionally cause OOM. So I make this change. Have you met this problem before?

File: example/rcnn/rcnn/warmup.py (hunk 2 above)
Comment (tornadomeet) [inline example/rcnn/rcnn/warmup.py]: @sxjscience how may gpu do you using when training? because the gpu number will influnce the batch-size, so it should be care of --factor-step(=5000 default).  my reproduce is 2gpu, and lr=0.001 in the fisrt 5 epoch, and lr=0.0001 in the last 2 epoch.

File: example/rcnn/rcnn/detector.py (hunk 3 above)
Comment (precedenceguo) [inline example/rcnn/rcnn/detector.py]: Yes that is due to gpu memory pool for efficient gpu memory usage. I think set MXNET_GPU_MEM_POOL_RESERVE=30 would address any OOM.
We could get an executor with possible maximum data memory and resize it down. But I just thought gpu mem pool reserve works fine.......

**Comment by sxjscience** [comment]
I've fixed the case for RCNN and have removed the warm-up scheduler. The mAP on VOC2007 is 0.6976 after training for 10 epochs on a single GPU using `train_end2end.py`

**Comment by tornadomeet** [comment]
@sxjscience great! !would you also give the result to the https://github.com/dmlc/mxnet/blob/master/example/rcnn/README.md, make a table of this two result(alternate and end2end ), i have test the alternate model before, map is about 0.6954.

**Comment by sxjscience** [comment]
OK, should be something like this. I'll update later.

## Results
|                     | Training Data | Test Data  |  mAP  |
| ------------- |:-------------:|:-----:| -----:|
| FasterRCNN (alternate) | VOC 07 trainval | VOC 07 test | 69.54% |
| FasterRCNN (approx-end2end) | VOC 07 trainval  | VOC 07 test | 69.76% |

**Comment by tornadomeet** [comment]
@sxjscience @precedenceguo @winstywang LGTM

**Comment by precedenceguo** [comment]
Very decent change. Let's wait for checks to finish. @winstywang
